File: libraries/cyclestreets-fragments/src/main/java/net/cyclestreets/SettingsFragment.java
package net.cyclestreets;

import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.transition.Fade;
import android.support.transition.Slide;
import android.support.v7.preference.*;

import android.util.Log;
import android.view.Gravity;
import net.cyclestreets.fragments.R;
import net.cyclestreets.tiles.TileSource;
import net.cyclestreets.util.Logging;
import net.cyclestreets.util.MapPack;
import net.cyclestreets.util.MessageBox;

import java.util.ArrayList;
import java.util.List;

public class SettingsFragment extends PreferenceFragmentCompat implements SharedPreferences.OnSharedPreferenceChangeListener
{
  private static final String TAG = Logging.getTag(SettingsFragment.class);
  private static final String PREFERENCE_SCREEN_ARG = "preferenceScreenArg";

  @Override
  public void onCreate(Bundle savedInstance) {
    super.onCreate(savedInstance);

    setupMapStyles();
    setupMapFileList();

    setSummary(CycleStreetsPreferences.PREF_ROUTE_TYPE_KEY);
    setSummary(CycleStreetsPreferences.PREF_UNITS_KEY);
    setSummary(CycleStreetsPreferences.PREF_SPEED_KEY);
    setSummary(CycleStreetsPreferences.PREF_MAPSTYLE_KEY);
    setSummary(CycleStreetsPreferences.PREF_MAPFILE_KEY);
    setSummary(CycleStreetsPreferences.PREF_UPLOAD_SIZE);
    setSummary(CycleStreetsPreferences.PREF_NEARING_TURN);
    setSummary(CycleStreetsPreferences.PREF_OFFTRACK_DISTANCE);
    setSummary(CycleStreetsPreferences.PREF_REPLAN_DISTANCE);
  }

  @Override
  public void onCreatePreferences(final Bundle savedInstance, final String rootKey) {
    if (getArguments() != null) {
      Log.d(TAG, "Creating preferences subscreen with key " + rootKey);
      String key = getArguments().getString(PREFERENCE_SCREEN_ARG);
      setPreferencesFromResource(R.xml.prefs, key);
    } else {
      Log.d(TAG, "Creating root preferences page");
      setPreferencesFromResource(R.xml.prefs, rootKey);
    }
  }

  @Override
  public void onNavigateToScreen(PreferenceScreen preferenceScreen) {
    // Allow instantiation of subscreens
    Bundle args = new Bundle();
    args.putString(PREFERENCE_SCREEN_ARG, preferenceScreen.getKey());
    SettingsFragment settingsSubScreen = new SettingsFragment();
    settingsSubScreen.setArguments(args);

    // Make the transitions into and back out of subscreens look nice
    settingsSubScreen.setEnterTransition(new Slide(Gravity.END));
    settingsSubScreen.setExitTransition(new Slide(Gravity.START));
    this.setEnterTransition(new Fade());
    this.setExitTransition(new Fade());

    getFragmentManager()
        .beginTransaction()
        .replace(getId(), settingsSubScreen)
        .addToBackStack(null)
        .commit();
  }

  private void setupMapStyles() {
    final ListPreference mapstylePref = (ListPreference)findPreference(CycleStreetsPreferences.PREF_MAPSTYLE_KEY);
    if (mapstylePref == null)
      return;
    TileSource.configurePreference(mapstylePref);
  }

  private void setupMapFileList() {
    final ListPreference mapfilePref = (ListPreference)findPreference(CycleStreetsPreferences.PREF_MAPFILE_KEY);
    if (mapfilePref == null)
      return;
    populateMapFileList(mapfilePref);
  }

  private void populateMapFileList(final ListPreference mapfilePref) {
    final List<String> names = new ArrayList<>();
    final List<String> files = new ArrayList<>();

    for (final MapPack pack : MapPack.availableMapPacks()) {
      names.add(pack.name());
      files.add(pack.path());
    }

    mapfilePref.setEntries(names.toArray(new String[] { }));
    mapfilePref.setEntryValues(files.toArray(new String[] { }));
  }

  @Override
  public void onResume() {
    super.onResume();

    getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);

    setAccountSummary();
  }

  @Override
  public void onPause() {
    super.onPause();

    // stop listening while paused
    getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
  }

  public void onSharedPreferenceChanged(final SharedPreferences prefs, final String key) {
    setSummary(key);
  }

  private void setSummary(final String key) {
    final Preference prefUI = findPreference(key);
    if (prefUI == null)
      return;
    if (prefUI instanceof ListPreference)
      prefUI.setSummary(((ListPreference)prefUI).getEntry());
    if (prefUI instanceof EditTextPreference)
      prefUI.setSummary(((EditTextPreference)prefUI).getText());

    if (CycleStreetsPreferences.PREF_MAPSTYLE_KEY.equals(key))
      setMapFileSummary(((ListPreference)prefUI).getValue());
  }

  private void setMapFileSummary(final String style) {
    final ListPreference mapfilePref = (ListPreference)findPreference(CycleStreetsPreferences.PREF_MAPFILE_KEY);
    if (mapfilePref == null)
      return;

    final boolean enabled = style.equals(CycleStreetsPreferences.MAPSTYLE_MAPSFORGE);
    mapfilePref.setEnabled(enabled);

    if (!enabled)
      return;

    if (mapfilePref.getEntryValues().length == 0) {
      mapfilePref.setEnabled(false);
      MessageBox.YesNo(getView(),
          R.string.settings_no_map_packs,
          (DialogInterface di, int i) -> MapPack.searchGooglePlay(getContext())
      );
      return;
    }

    final String mapfile = CycleStreetsPreferences.mapfile();
    int index = mapfilePref.findIndexOfValue(mapfile);
    if (index == -1)
      index = 0; // default to something

    mapfilePref.setValueIndex(index);
    mapfilePref.setSummary(mapfilePref.getEntries()[index]);
  }

  private void setAccountSummary() {
    final PreferenceScreen account = (PreferenceScreen)findPreference(CycleStreetsPreferences.PREF_ACCOUNT_KEY);
    if (account == null)
      return;

    if (CycleStreetsPreferences.accountOK())
      account.setSummary(R.string.settings_signed_in);
    else if (CycleStreetsPreferences.accountPending())
      account.setSummary(R.string.settings_awaiting);
    else
      account.setSummary("");
  }
}

File: libraries/cyclestreets-fragments/src/main/java/net/cyclestreets/SettingsFragment.kt
package net.cyclestreets

import android.content.SharedPreferences
import android.os.Bundle
import androidx.transition.Fade
import androidx.transition.Slide
import androidx.preference.*

import android.util.Log
import android.view.Gravity
import com.mikepenz.iconics.typeface.library.googlematerial.GoogleMaterial
import net.cyclestreets.fragments.R
import net.cyclestreets.iconics.IconicsHelper.materialIcon
import net.cyclestreets.tiles.TileSource
import net.cyclestreets.util.Logging
import net.cyclestreets.util.MapPack
import net.cyclestreets.util.MessageBox


private val TAG = Logging.getTag(SettingsFragment::class.java)
private const val PREFERENCE_SCREEN_ARG: String = "preferenceScreenArg"
private val SETTINGS_ICONS = mapOf(
    "screen-maps-display" to GoogleMaterial.Icon.gmd_map,
    "mapstyle" to null,
    "confirm-new-route" to null,
    "screen-routing-preferences" to GoogleMaterial.Icon.gmd_directions,
    "routetype" to null,
    "speed" to null,
    "units" to null,
    "screen-liveride" to GoogleMaterial.Icon.gmd_navigation,
    "nearing-turn-distance" to null,
    "offtrack-distance" to null,
    "replan-distance" to null,
    "screen-locations" to GoogleMaterial.Icon.gmd_edit_location,
    "screen-account" to GoogleMaterial.Icon.gmd_account_circle,
    "cyclestreets-account" to null,
    "username" to null,
    "password" to null,
    "uploadsize" to null,
    "screen-about" to GoogleMaterial.Icon.gmd_info_outline
)


class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener, Undoable {

    private var undoable = false

    override fun onCreate(savedInstance: Bundle?) {
        super.onCreate(savedInstance)

        setupMapStyles()

        setSummary(CycleStreetsPreferences.PREF_ROUTE_TYPE_KEY)
        setSummary(CycleStreetsPreferences.PREF_UNITS_KEY)
        setSummary(CycleStreetsPreferences.PREF_SPEED_KEY)
        setSummary(CycleStreetsPreferences.PREF_MAPSTYLE_KEY)
        setSummary(CycleStreetsPreferences.PREF_UPLOAD_SIZE)
        setSummary(CycleStreetsPreferences.PREF_NEARING_TURN)
        setSummary(CycleStreetsPreferences.PREF_OFFTRACK_DISTANCE)
        setSummary(CycleStreetsPreferences.PREF_REPLAN_DISTANCE)
    }

    override fun onCreatePreferences(savedInstance: Bundle?, rootKey: String?) {
        if (arguments != null) {
            val key = requireArguments().getString(PREFERENCE_SCREEN_ARG)
            Log.d(TAG, "Creating preferences subscreen with key $key")
            setPreferencesFromResource(R.xml.prefs, key)
            undoable = true
            this.enterTransition = Slide(Gravity.END)
            this.exitTransition = Slide(Gravity.END)
        } else {
            Log.d(TAG, "Creating root preferences page")
            setPreferencesFromResource(R.xml.prefs, rootKey)
            this.enterTransition = Fade()
            this.exitTransition = Fade()
        }
        populateSettingsIcons();
    }

    private fun populateSettingsIcons() {
        for (prefIndex in 0 until preferenceScreen.preferenceCount) {
            val pref = preferenceScreen.getPreference(prefIndex)
            SETTINGS_ICONS[pref.key]?.let {
                pref.icon = materialIcon(requireContext(), it)
            }
        }
    }

    override fun onNavigateToScreen(preferenceScreen: PreferenceScreen) {
        showScreen(preferenceScreen.key)
    }

    override fun onBackPressed(): Boolean {
        if (undoable)
            showScreen()
        return undoable
    }

    private fun showScreen(key: String? = null) {
        val screen = SettingsFragment()

        if (key != null) {
            val args = Bundle()
            args.putString(PREFERENCE_SCREEN_ARG, key)
            screen.arguments = args
        }

        parentFragmentManager
                .beginTransaction()
                .replace(id, screen)
                .commit()

    }

    private fun setupMapStyles() {
        findPreference<ListPreference>(CycleStreetsPreferences.PREF_MAPSTYLE_KEY)?.apply {
           TileSource.configurePreference(this)
        }
    }

    override fun onResume() {
        super.onResume()

        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(this)

        setAccountSummary()
    }

    override fun onPause() {
        super.onPause()

        // stop listening while paused
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onSharedPreferenceChanged(prefs: SharedPreferences, key: String) {
        setSummary(key)
    }

    private fun setSummary(key: String) {
        val prefUI = findPreference<Preference>(key) ?: return
        if (prefUI is ListPreference)
            prefUI.summary = prefUI.entry
        if (prefUI is EditTextPreference)
            prefUI.summary = prefUI.text
   }

   private fun setAccountSummary() {
        val pref = findPreference<Preference>(CycleStreetsPreferences.PREF_ACCOUNT_KEY) ?: return
        val account = pref as PreferenceScreen

        when {
            CycleStreetsPreferences.accountOK() -> account.setSummary(R.string.settings_signed_in)
            CycleStreetsPreferences.accountPending() -> account.setSummary(R.string.settings_awaiting)
            else -> account.summary = ""
        }
    }
}
